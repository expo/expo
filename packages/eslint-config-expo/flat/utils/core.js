const expo = require('eslint-plugin-expo');
const importPlugin = require('eslint-plugin-import');
const { jsExtensions } = require('./extensions');

const globals = require('globals');

module.exports = [
  importPlugin.flatConfigs.recommended,
  importPlugin.flatConfigs.errors,
  {
    plugins: {
      expo,
    },

    languageOptions: {
      globals: {
        console: 'readonly',
        exports: false,
        global: false,
        module: false,
        require: false,
      },

      ecmaVersion: 2022,
      sourceType: 'module',

      parserOptions: {
        ecmaFeatures: {
          impliedStrict: true,
          jsx: true,
        },
      },
    },

    settings: {
      'import/ignore': ['node_modules[\\\\/]+@?react-native'],

      'import/extensions': jsExtensions,
      'import/resolver': {
        node: { extensions: jsExtensions },
        typescript: true,
      },
    },

    rules: {
      eqeqeq: ['warn', 'smart'],
      'no-dupe-args': 'error',
      'no-dupe-class-members': 'error',
      'no-dupe-keys': 'error',
      'no-duplicate-case': 'error',
      'no-empty-character-class': 'warn',
      'no-empty-pattern': 'warn',
      'no-extend-native': 'warn',
      'no-extra-bind': 'warn',
      'no-redeclare': 'warn',
      'no-undef': 'error',
      'no-unreachable': 'warn',
      'no-unsafe-negation': 'warn',

      'no-unused-expressions': [
        'warn',
        {
          allowShortCircuit: true,
          enforceForJSX: true,
        },
      ],

      'no-unused-labels': 'warn',

      'no-unused-vars': [
        'warn',
        {
          vars: 'all',
          args: 'none',
          ignoreRestSiblings: true,
          caughtErrors: 'all',
          caughtErrorsIgnorePattern: '^_',
        },
      ],

      'no-with': 'warn',
      'unicode-bom': ['warn', 'never'],
      'use-isnan': 'error',
      'valid-typeof': 'error',
      'import/first': 'warn',
      'import/default': 'off',
    },
  },
  {
    files: ['**/*.d.ts'],

    rules: {
      'import/order': 'off',
    },
  },
  {
    files: ['**/metro.config.js'],

    languageOptions: {
      globals: globals.node,
    },
  },
];
