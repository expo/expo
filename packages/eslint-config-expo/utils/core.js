const { jsExtensions } = require('./extensions');

module.exports = {
  parserOptions: {
    sourceType: 'module',
    ecmaVersion: 2022,
    ecmaFeatures: { impliedStrict: true, jsx: true },
  },
  env: { es2022: true },
  globals: {
    console: 'readonly',
    exports: false,
    global: false,
    module: false,
    require: false,
  },
  extends: ['plugin:import/errors'],
  plugins: ['import', 'expo'],
  rules: {
    eqeqeq: ['warn', 'smart'],
    'no-dupe-args': 'error',
    'no-dupe-class-members': 'error',
    'no-dupe-keys': 'error',
    'no-duplicate-case': 'error',
    'no-empty-character-class': 'warn',
    'no-empty-pattern': 'warn',
    'no-extend-native': 'warn',
    'no-extra-bind': 'warn',
    'no-redeclare': 'warn',
    'no-undef': 'error',
    'no-unreachable': 'warn',
    'no-unsafe-negation': 'warn',
    'no-unused-expressions': ['warn', { allowShortCircuit: true, enforceForJSX: true }],
    'no-unused-labels': 'warn',
    'no-unused-vars': [
      'warn',
      {
        vars: 'all',
        args: 'none',
        ignoreRestSiblings: true,
        caughtErrors: 'all',
        caughtErrorsIgnorePattern: '^_',
      },
    ],
    'no-with': 'warn',
    'unicode-bom': ['warn', 'never'],
    'use-isnan': 'error',
    'valid-typeof': 'error',
    'import/first': 'warn', //keep
    'import/default': 'off',
  },
  settings: {
    'import/extensions': jsExtensions,
    'import/ignore': [
      // react-native's main module is Flow, not JavaScript, and raises parse errors. Additionally,
      // several other react-native-related packages still publish Flow code as their main source.
      'node_modules[\\\\/]+@?react-native',
    ],
    'import/resolver': {
      node: { extensions: jsExtensions },
      typescript: true
    },
  },
  overrides: [
    {
      files: ['*.d.ts'],
      rules: {
        'import/order': 'off',
      },
    },
  ],
};
