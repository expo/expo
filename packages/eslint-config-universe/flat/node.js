const { defineConfig } = require('eslint/config');
const nodePlugin = require('eslint-plugin-n');
const globals = require('globals');

const coreConfig = require('./shared/core.js');
const prettierConfig = require('./shared/prettier.js');
const typescriptConfig = require('./shared/typescript.js');

module.exports = defineConfig([
  {
    plugins: {
      n: nodePlugin,
    },

    languageOptions: {
      globals: {
        ...globals.node,
      },
    },

    rules: {
      'n/no-deprecated-api': 'warn',
      'n/no-path-concat': 'warn',
    },
  },
  coreConfig,
  typescriptConfig,
  prettierConfig,
]);
