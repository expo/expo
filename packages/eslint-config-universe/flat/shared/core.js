const { defineConfig } = require('eslint/config');
const importPlugin = require('eslint-plugin-import');
const nodePlugin = require('eslint-plugin-n');
const globals = require('globals');

const { jsExtensions } = require('./extensions');

module.exports = defineConfig([
  importPlugin.flatConfigs.recommended,
  {
    plugins: {
      n: nodePlugin,
    },

    languageOptions: {
      globals: {
        ...globals.jest,
        console: false,
        exports: false,
        global: false,
        module: false,
        require: false,
      },

      ecmaVersion: 2022,
      sourceType: 'module',

      parserOptions: {
        ecmaFeatures: {
          impliedStrict: true,
          jsx: true,
        },
      },
    },

    settings: {
      'import/extensions': jsExtensions,
      'import/ignore': ['node_modules[\\\\/]+@?react-native'],

      'import/resolver': {
        node: {
          extensions: jsExtensions,
        },
      },
    },

    rules: {
      'array-bracket-spacing': ['warn', 'never'],

      'arrow-spacing': [
        'warn',
        {
          before: true,
          after: true,
        },
      ],

      curly: ['warn', 'all'],
      'block-spacing': ['warn', 'always'],
      'comma-dangle': ['warn', 'always-multiline'],

      'comma-spacing': [
        'warn',
        {
          before: false,
          after: true,
        },
      ],

      'comma-style': ['warn', 'last'],
      'computed-property-spacing': ['warn', 'never'],
      'constructor-super': 'warn',
      'dot-location': ['warn', 'property'],
      'eol-last': 'warn',
      eqeqeq: ['warn', 'smart'],
      'func-call-spacing': ['warn', 'never'],
      'generator-star-spacing': ['warn', 'after'],
      'getter-return': 'warn',
      'jsx-quotes': ['warn', 'prefer-double'],
      'new-parens': 'warn',
      'no-alert': 'off',
      'no-array-constructor': 'warn',
      'no-caller': 'warn',
      'no-case-declarations': 'warn',
      'no-compare-neg-zero': 'warn',
      'no-cond-assign': 'warn',
      'no-const-assign': 'error',

      'no-constant-condition': [
        'warn',
        {
          checkLoops: false,
        },
      ],

      'no-control-regex': 'off',
      'no-debugger': 'warn',
      'no-delete-var': 'error',
      'no-dupe-args': 'error',
      'no-dupe-class-members': 'error',
      'no-dupe-keys': 'error',
      'no-duplicate-case': 'error',
      'no-empty-character-class': 'warn',
      'no-empty-pattern': 'warn',
      'no-eval': 'warn',
      'no-ex-assign': 'warn',
      'no-extend-native': 'warn',
      'no-extra-bind': 'warn',
      'no-extra-boolean-cast': 'warn',
      'no-extra-semi': 'warn',
      'no-fallthrough': 'warn',
      'no-floating-decimal': 'warn',
      'no-func-assign': 'error',
      'no-global-assign': 'warn',
      'no-implied-eval': 'warn',
      'no-inner-declarations': 'warn',
      'no-invalid-regexp': 'error',
      'no-irregular-whitespace': 'warn',
      'no-iterator': 'warn',
      'no-label-var': 'warn',

      'no-labels': [
        'warn',
        {
          allowLoop: true,
          allowSwitch: true,
        },
      ],

      'no-lone-blocks': 'warn',
      'no-multi-assign': 'warn',
      'no-new': 'warn',
      'no-new-func': 'warn',
      'no-object-constructor': 'warn',
      'no-new-native-nonconstructor': 'error',
      'no-obj-calls': 'warn',
      'no-octal': 'warn',
      'no-octal-escape': 'warn',
      'no-mixed-spaces-and-tabs': 'warn',
      'no-proto': 'warn',
      'no-redeclare': 'warn',
      'no-return-assign': 'warn',
      'no-script-url': 'warn',
      'no-self-assign': 'warn',
      'no-self-compare': 'warn',
      'no-sequences': 'warn',
      'no-shadow-restricted-names': 'warn',
      'no-sparse-arrays': 'warn',
      'no-this-before-super': 'warn',
      'no-throw-literal': 'warn',
      'no-trailing-spaces': 'warn',
      'no-undef': 'error',
      'no-unexpected-multiline': 'warn',
      'no-unneeded-ternary': 'warn',
      'no-unreachable': 'warn',
      'no-unsafe-negation': 'warn',

      'no-unused-expressions': [
        'warn',
        {
          allowShortCircuit: true,
          enforceForJSX: true,
        },
      ],

      'no-unused-labels': 'warn',

      'no-unused-vars': [
        'warn',
        {
          vars: 'all',
          args: 'none',
          ignoreRestSiblings: true,
          caughtErrors: 'all',
          caughtErrorsIgnorePattern: '^_',
        },
      ],

      'no-useless-computed-key': 'warn',
      'no-useless-concat': 'warn',
      'no-useless-constructor': 'warn',
      'no-useless-escape': 'warn',
      'no-useless-rename': 'warn',
      'no-useless-return': 'warn',
      'no-var': 'warn',
      'no-void': 'warn',
      'no-whitespace-before-property': 'warn',
      'no-with': 'warn',
      'object-shorthand': 'warn',

      'operator-linebreak': [
        'warn',
        'after',
        {
          overrides: {
            '?': 'before',
            ':': 'before',
          },
        },
      ],

      'prefer-const': [
        'warn',
        {
          destructuring: 'all',
          ignoreReadBeforeAssign: true,
        },
      ],

      'prefer-promise-reject-errors': 'warn',
      'prefer-rest-params': 'warn',
      'prefer-spread': 'warn',
      radix: 'warn',
      'rest-spread-spacing': ['warn', 'never'],
      semi: 'warn',

      'semi-spacing': [
        'warn',
        {
          before: false,
          after: true,
        },
      ],

      'semi-style': ['warn', 'last'],
      'space-before-blocks': ['warn', 'always'],

      'space-before-function-paren': [
        'warn',
        {
          anonymous: 'never',
          named: 'never',
        },
      ],

      'space-in-parens': ['warn', 'never'],
      'space-infix-ops': 'warn',

      'switch-colon-spacing': [
        'warn',
        {
          before: false,
          after: true,
        },
      ],

      'template-curly-spacing': ['warn', 'never'],
      'template-tag-spacing': ['warn', 'never'],
      'unicode-bom': ['warn', 'never'],
      'use-isnan': 'error',
      'valid-typeof': 'error',
      'yield-star-spacing': ['warn', 'after'],

      yoda: [
        'warn',
        'never',
        {
          exceptRange: true,
        },
      ],

      'import/default': 'off',
      'import/export': 'error',
      'import/first': 'warn',

      'import/namespace': [
        'error',
        {
          allowComputed: true,
        },
      ],

      'import/no-duplicates': 'error',

      // TODO (Kadi): Enable this. Disabling for now because import/recommended adds it, but we didn't use to have it enabled
      'import/no-unresolved': 'off',

      'import/order': [
        'warn',
        {
          groups: [['builtin', 'external'], 'internal', ['parent', 'index', 'sibling']],
          'newlines-between': 'always',

          alphabetize: {
            order: 'asc',
          },
        },
      ],

      'n/handle-callback-err': ['warn', '^(e|err|error|.+Error)$'],
      'n/no-new-require': 'warn',
    },
  },
  {
    files: ['**/*.d.ts'],

    rules: {
      'import/order': 'off',
    },
  },
]);
