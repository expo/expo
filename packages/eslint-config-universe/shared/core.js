const { jsExtensions } = require('./extensions');

module.exports = {
  parserOptions: {
    sourceType: 'module',
    ecmaVersion: 2022,
    ecmaFeatures: { impliedStrict: true, jsx: true },
  },
  env: { es2022: true, jest: true },
  globals: {
    console: false,
    exports: false,
    global: false,
    module: false,
    require: false,
  },
  plugins: ['import', 'node'],
  rules: {
    'array-bracket-spacing': ['warn', 'never'],
    'arrow-spacing': ['warn', { before: true, after: true }],
    curly: ['warn', 'all'],
    'block-spacing': ['warn', 'always'],
    'comma-dangle': ['warn', 'always-multiline'],
    'comma-spacing': ['warn', { before: false, after: true }],
    'comma-style': ['warn', 'last'],
    'computed-property-spacing': ['warn', 'never'],
    'constructor-super': 'warn',
    'dot-location': ['warn', 'property'],
    'eol-last': 'warn',
    eqeqeq: ['warn', 'smart'],
    'func-call-spacing': ['warn', 'never'],
    'generator-star-spacing': ['warn', 'after'],
    'getter-return': 'warn',
    'jsx-quotes': ['warn', 'prefer-double'],
    'new-parens': 'warn',
    'no-alert': 'off',
    'no-array-constructor': 'warn',
    'no-caller': 'warn',
    'no-case-declarations': 'warn',
    'no-compare-neg-zero': 'warn',
    'no-cond-assign': 'warn',
    'no-const-assign': 'error',
    'no-constant-condition': ['warn', { checkLoops: false }],
    'no-control-regex': 'off',
    'no-debugger': 'warn',
    'no-delete-var': 'error',
    'no-dupe-args': 'error',
    'no-dupe-class-members': 'error',
    'no-dupe-keys': 'error',
    'no-duplicate-case': 'error',
    'no-empty-character-class': 'warn',
    'no-empty-pattern': 'warn',
    'no-eval': 'warn',
    'no-ex-assign': 'warn',
    'no-extend-native': 'warn',
    'no-extra-bind': 'warn',
    'no-extra-boolean-cast': 'warn',
    'no-extra-semi': 'warn',
    'no-fallthrough': 'warn',
    'no-floating-decimal': 'warn',
    'no-func-assign': 'error',
    'no-global-assign': 'warn',
    'no-implied-eval': 'warn',
    'no-inner-declarations': 'warn',
    'no-invalid-regexp': 'error',
    'no-irregular-whitespace': 'warn',
    'no-iterator': 'warn',
    'no-label-var': 'warn',
    'no-labels': ['warn', { allowLoop: true, allowSwitch: true }],
    'no-lone-blocks': 'warn',
    'no-multi-assign': 'warn',
    'no-new': 'warn',
    'no-new-func': 'warn',
    'no-new-object': 'warn',
    'no-new-symbol': 'error',
    'no-obj-calls': 'warn',
    'no-octal': 'warn',
    'no-octal-escape': 'warn',
    'no-mixed-spaces-and-tabs': 'warn',
    'no-proto': 'warn',
    'no-redeclare': 'warn',
    'no-return-assign': 'warn',
    'no-script-url': 'warn',
    'no-self-assign': 'warn',
    'no-self-compare': 'warn',
    'no-sequences': 'warn',
    'no-shadow-restricted-names': 'warn',
    'no-sparse-arrays': 'warn',
    'no-this-before-super': 'warn',
    'no-throw-literal': 'warn',
    'no-trailing-spaces': 'warn',
    'no-undef': 'error',
    'no-unexpected-multiline': 'warn',
    'no-unneeded-ternary': 'warn',
    'no-unreachable': 'warn',
    'no-unsafe-negation': 'warn',
    'no-unused-expressions': ['warn', { allowShortCircuit: true, enforceForJSX: true }],
    'no-unused-labels': 'warn',
    'no-unused-vars': [
      'warn',
      {
        vars: 'all',
        args: 'none',
        ignoreRestSiblings: true,
        caughtErrors: 'all',
        caughtErrorsIgnorePattern: '^_',
      },
    ],
    'no-useless-computed-key': 'warn',
    'no-useless-concat': 'warn',
    'no-useless-constructor': 'warn',
    'no-useless-escape': 'warn',
    'no-useless-rename': 'warn',
    'no-useless-return': 'warn',
    'no-var': 'warn',
    'no-void': 'warn',
    'no-whitespace-before-property': 'warn',
    'no-with': 'warn',
    'object-shorthand': 'warn',
    'operator-linebreak': ['warn', 'after', { overrides: { '?': 'before', ':': 'before' } }],
    'prefer-const': ['warn', { destructuring: 'all', ignoreReadBeforeAssign: true }],
    'prefer-promise-reject-errors': 'warn',
    'prefer-rest-params': 'warn',
    'prefer-spread': 'warn',
    radix: 'warn',
    'rest-spread-spacing': ['warn', 'never'],
    semi: 'warn',
    'semi-spacing': ['warn', { before: false, after: true }],
    'semi-style': ['warn', 'last'],
    'space-before-blocks': ['warn', 'always'],
    'space-before-function-paren': ['warn', { anonymous: 'never', named: 'never' }],
    'space-in-parens': ['warn', 'never'],
    'space-infix-ops': 'warn',
    'switch-colon-spacing': ['warn', { before: false, after: true }],
    'template-curly-spacing': ['warn', 'never'],
    'template-tag-spacing': ['warn', 'never'],
    'unicode-bom': ['warn', 'never'],
    'use-isnan': 'error',
    'valid-typeof': 'error',
    'yield-star-spacing': ['warn', 'after'],
    yoda: ['warn', 'never', { exceptRange: true }],

    'import/default': 'off',
    'import/export': 'error',
    'import/first': 'warn',
    'import/namespace': ['error', { allowComputed: true }],
    'import/no-duplicates': 'error',
    'import/order': [
      'warn',
      {
        groups: [['builtin', 'external'], 'internal', ['parent', 'index', 'sibling']],
        'newlines-between': 'always',
        alphabetize: {
          order: 'asc',
        },
      },
    ],

    'node/handle-callback-err': ['warn', '^(e|err|error|.+Error)$'],
    'node/no-new-require': 'warn',
  },
  settings: {
    'import/extensions': jsExtensions,
    'import/ignore': [
      // react-native's main module is Flow, not JavaScript, and raises parse errors. Additionally,
      // several other react-native-related packages still publish Flow code as their main source.
      'node_modules[\\\\/]+@?react-native',
    ],
    'import/resolver': {
      node: { extensions: jsExtensions },
    },
  },
  overrides: [
    {
      files: ['*.d.ts'],
      rules: {
        'import/order': 'off',
      },
    },
  ],
};
