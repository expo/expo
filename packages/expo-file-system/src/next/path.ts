export function join(...paths: string[]) {
  return paths.join('/');
}

export const separator = '/';

export function dirname(path: string) {
  if (path.endsWith('/')) {
    return path;
  }
  return `${path.split('/').slice(0, -1).join('/')}/`;
}

export function basename(path: string, suffix?: string) {
  const pathWithoutTrailingSeparator = path.endsWith('/') ? path.slice(0, -1) : path;
  const basename = pathWithoutTrailingSeparator.split('/').pop() || '';
  if (!suffix) {
    return basename;
  }
  if (basename.endsWith(suffix)) {
    return basename.slice(0, -suffix.length);
  }
  return basename;
}

export function extname(path: string) {
  const base = basename(path);
  const lastDotIndex = base.lastIndexOf('.');
  if (lastDotIndex === -1) {
    return '';
  }
  return base.slice(lastDotIndex + 1);
}
