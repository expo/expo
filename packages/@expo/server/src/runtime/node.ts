import fs from 'node:fs';
import path from 'node:path';

import { RawManifest, Manifest, Route, Middleware } from '../types';
import { initManifestRegExp } from '../utils/initManifestRegExp';

const debug =
  process.env.NODE_ENV === 'development'
    ? (require('debug')('expo:server') as typeof console.log)
    : () => {};

export const handleRouteError = () => async (error: Error) => {
  throw error;
};

export const getRoutesManifest = (dist: string) => async (): Promise<Manifest> => {
  const raw = path.join(dist, '_expo/routes.json');
  // TODO: JSON Schema for validation
  const manifest = JSON.parse(fs.readFileSync(raw, 'utf-8')) as RawManifest;
  return initManifestRegExp(manifest);
};

export const getHtml =
  (dist: string) =>
  async (_request: Request, route: Route): Promise<string | null> => {
    // Serve a static file by exact route name
    const filePath = path.join(dist, route.page + '.html');
    if (fs.existsSync(filePath)) {
      return fs.readFileSync(filePath, 'utf-8');
    }

    // Serve a static file by route name with hoisted index
    // See: https://github.com/expo/expo/pull/27935
    const hoistedFilePath = route.page.match(/\/index$/)
      ? path.join(dist, route.page.replace(/\/index$/, '') + '.html')
      : null;
    if (hoistedFilePath && fs.existsSync(hoistedFilePath)) {
      return fs.readFileSync(hoistedFilePath, 'utf-8');
    }

    return null;
  };

export const getApiRoute = (dist: string) => async (route: Route) => {
  debug(`Handling API route: ${route.page}: ${route.file}`);
  return loadServerModule(dist, route);
};

export const getMiddleware = (dist: string) => async (middleware: Middleware) => {
  return loadServerModule(dist, middleware);
};

async function loadServerModule(dist: string, mod: { file: string }): Promise<any> {
  const filePath = path.join(dist, mod.file);

  // TODO: What's the standard behavior for malformed projects?
  if (!fs.existsSync(filePath)) {
    return null;
  }

  if (/\.c?js$/.test(filePath)) {
    return require(filePath);
  }
  return import(filePath);
}
