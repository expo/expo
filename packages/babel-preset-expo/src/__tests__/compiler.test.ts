import * as babel from '@babel/core';
import * as path from 'node:path';

import preset from '..';

function getCaller(props: Record<string, string | boolean>): babel.TransformCaller {
  return props as unknown as babel.TransformCaller;
}

const options: babel.TransformOptions = {
  sourceType: 'unambiguous',
  caller: getCaller({
    name: 'metro',
    supportsReactCompiler: true,
    engine: 'hermes',
    platform: 'ios',
    isDev: false,
  }),
  babelrc: false,
  presets: [preset],
  sourceMaps: false,

  compact: false,
  comments: true,
  retainLines: true,
};

it(`allows destructuring in the catch block`, () => {
  // Ensuring the transform doesn't throw.
  const { code } = babel.transformFileSync(
    path.resolve(__dirname, 'samples/destructure-catch.tsx'),
    options
  )!;

  expect(code).toMatchSnapshot();
  expect(code).not.toContain('react.memo_cache_sentinel');
});

it(`supports functions with discarded return values in try/catch blocks that run in memos`, () => {
  // Ensuring the transform doesn't throw.
  const { code } = babel.transformFileSync(
    path.resolve(__dirname, 'samples/try-catch-hook.tsx'),
    options
  )!;

  expect(code).toMatchSnapshot();
  expect(code).not.toContain('react.memo_cache_sentinel');
});

it(`supports memoizing`, () => {
  // Ensuring the transform doesn't throw.
  const { code } = babel.transformFileSync(path.resolve(__dirname, 'samples/App.tsx'), options)!;

  expect(code).toMatchSnapshot();
  expect(code).toContain('react.memo_cache_sentinel');
});

it(`supports disabling memoizing`, () => {
  // Ensuring the transform doesn't throw.
  const { code } = babel.transformFileSync(path.resolve(__dirname, 'samples/App.tsx'), {
    ...options,
    caller: getCaller({
      name: 'metro',
      supportsReactCompiler: false,
      engine: 'hermes',
      platform: 'ios',
      isDev: false,
    }),
  })!;

  expect(code).not.toContain('react.memo_cache_sentinel');
});

it(`compiles to CJS 1`, () => {
  const code = babel.transformFileSync(path.resolve(__dirname, 'samples/compile-memo.tsx'), {
    ...options,
    caller: getCaller({
      name: 'metro',
      supportsReactCompiler: true,
      supportsStaticESM: false,
      engine: 'hermes',
      platform: 'ios',
      isDev: true,
    }),
  })!.code!;
  expect(code).not.toContain('import {');
  expect(code).toContain('react.memo_cache_sentinel');
});

// TODO: Known issue with compiler.
xit(`compiles React module with js extension in filename`, () => {
  const code = babel.transformSync(
    `
// import React from 'react';
// const React2 = require('react');
function App() {
  return <div />;
}
`,
    {
      ...options,
      // File name ending with `.js`
      filename: '/samples/App.js',
      // This is the default type used in Metro.
      sourceType: 'unambiguous',
      babelrc: false,
      comments: false,
      code: true,
      caller: getCaller({
        name: 'metro',
        platform: 'web',
        isServer: false,
        isReactServer: false,
        isDev: true,
        isNodeModule: false,
        isHMREnabled: true,
        supportsStaticESM: false,
        supportsReactCompiler: true,
      }),
    }
  )!.code!;
  expect(code).toContain('var _reactCompilerRuntime = ');
  expect(code).not.toContain('import ');
});

it(`skips memoizing in server bundling passes`, () => {
  const { code } = babel.transformFileSync(path.resolve(__dirname, 'samples/PureComponent.tsx'), {
    ...options,
    caller: getCaller({
      isServer: true,
      name: 'metro',
      supportsReactCompiler: true,
      engine: 'hermes',
      platform: 'ios',
      isDev: false,
    }),
  })!;
  expect(code).not.toContain('react.memo_cache_sentinel');
});

it(`skips memoizing in react-server bundling passes`, () => {
  const { code } = babel.transformFileSync(path.resolve(__dirname, 'samples/PureComponent.tsx'), {
    ...options,
    caller: getCaller({
      isReactServer: true,
      name: 'metro',
      supportsReactCompiler: true,
      engine: 'hermes',
      platform: 'ios',
      isDev: false,
    }),
  })!;
  expect(code).not.toContain('react.memo_cache_sentinel');
});

it(`skips memoizing in node modules`, () => {
  const { code } = babel.transformFileSync(path.resolve(__dirname, 'samples/PureComponent.tsx'), {
    ...options,
    caller: getCaller({
      isNodeModule: true,
      name: 'metro',
      supportsReactCompiler: true,
      engine: 'hermes',
      platform: 'ios',
      isDev: false,
    }),
  })!;
  expect(code).not.toContain('react.memo_cache_sentinel');
});
