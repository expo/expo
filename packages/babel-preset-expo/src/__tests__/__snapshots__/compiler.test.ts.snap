// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allows destructuring in the catch block 1`] = `
"Object.defineProperty(exports, "__esModule", { value: true });exports.App = App;var _jsxRuntime = require("react/jsx-runtime");function App() {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  var foo = () => {
    try {
      console.log('hey');
    } catch (_ref) {var message = _ref.message;
      // NOTE(EvanBacon): React compiler throws:
      // (BuildHIR::lowerAssignment) Could not find binding for declaration.
      console.error(message);
    }
  };
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", { children: "Hello" });
}"
`;

exports[`supports functions with discarded return values in try/catch blocks that run in memos 1`] = `
"Object.defineProperty(exports, "__esModule", { value: true });exports.useSideEffectMayThrow = useSideEffectMayThrow;var _react = require("react");

function SIDE_EFFECT_MAY_THROW() {}

// NOTE(EvanBacon): React compiler throws:
// Cannot read properties of undefined (reading 'preds')
function useSideEffectMayThrow() {var t0;try {


    SIDE_EFFECT_MAY_THROW();
    t0 = true;} catch {

    t0 = false;}return t0;}"
`;

exports[`supports memoizing 1`] = `
"var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");Object.defineProperty(exports, "__esModule", { value: true });exports.default = App;var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));var _reactCompilerRuntime = require("react-compiler-runtime");var _vectorIcons = require("@expo/vector-icons");
var _expoAppLoading = _interopRequireDefault(require("expo-app-loading"));
var _expoAsset = require("expo-asset");
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");var _jsxRuntime = require("react/jsx-runtime");function _getRequireWildcardCache(e) {if ("function" != typeof WeakMap) return null;var r = new WeakMap(),t = new WeakMap();return (_getRequireWildcardCache = function (e) {return e ? t : r;})(e);}function _interopRequireWildcard(e, r) {if (!r && e && e.__esModule) return e;if (null === e || "object" != typeof e && "function" != typeof e) return { default: e };var t = _getRequireWildcardCache(r);if (t && t.has(e)) return t.get(e);var n = { __proto__: null },a = Object.defineProperty && Object.getOwnPropertyDescriptor;for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) {var i = a ? Object.getOwnPropertyDescriptor(e, u) : null;i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u];}return n.default = e, t && t.set(e, n), n;}

function App() {var $ = (0, _reactCompilerRuntime.c)(5);
  var _useState = (0, _react.useState)(false),_useState2 = (0, _slicedToArray2.default)(_useState, 2),assetsAreLoaded = _useState2[0],setAssetsAreLoaded = _useState2[1];var t0;var t1;if ($[0] === Symbol.for("react.memo_cache_sentinel")) {

    t0 = () => {
      loadAssetsAsync(setAssetsAreLoaded);};
    t1 = [];$[0] = t0;$[1] = t1;} else {t0 = $[0];t1 = $[1];}(0, _react.useEffect)(t0, t1);if (!

  assetsAreLoaded) {var _t;if ($[2] === Symbol.for("react.memo_cache_sentinel")) {
      _t = /*#__PURE__*/(0, _jsxRuntime.jsx)(_expoAppLoading.default, {});$[2] = _t;} else {_t = $[2];}return _t;}var t2;if ($[3] !== assetsAreLoaded) {


    t2 = !assetsAreLoaded ? /*#__PURE__*/
    (0, _jsxRuntime.jsx)(_expoAppLoading.default, {}) : /*#__PURE__*/

    (0, _jsxRuntime.jsx)(_reactNative.View, { style: styles.container, children: /*#__PURE__*/
      (0, _jsxRuntime.jsx)(_vectorIcons.Ionicons, { name: "md-options", size: 28 }) }
    );$[3] = assetsAreLoaded;$[4] = t2;} else {t2 = $[4];}return t2;}function



loadAssetsAsync(_x) {return _loadAssetsAsync.apply(this, arguments);}function _loadAssetsAsync() {_loadAssetsAsync = (0, _asyncToGenerator2.default)(function* (
  setAssetsAreLoaded)
  {
    try {
      yield _expoAsset.Asset.loadAsync([require('./assets/icon.png')]);
    } finally {
      setAssetsAreLoaded(true);
    }
  });return _loadAssetsAsync.apply(this, arguments);}

var styles = _reactNative.StyleSheet.create({
  container: {
    flex: 1
  }
});"
`;
